### PR TITLE
Query Frontend: Add tests and fix previous bugs for instant query tripperware

### PR DESCRIPTION
**Pull request by yeya24**
Signed-off-by: Ben Ye <ben.ye@bytedance.com>

Follow up PR for https://github.com/thanos-io/thanos/pull/5561.

* [ ] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

## Changes

1. Add the missing changelog.
2. Add more test cases and PromQL compatibility tests for query frontend with sharding.
3. Fix results to handle `string` and `matrix` for instant queries. Those types are not handled correctly in the previous pr.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -34,6 +34,7 @@ We use *breaking :warning:* to mark changes that are not backward compatible (re
 - [#5575](https://github.com/thanos-io/thanos/pull/5575) Receive: Add support for gRPC compression with snappy.
 - [#5508](https://github.com/thanos-io/thanos/pull/5508) Receive: Validate labels in write requests.
 - [#5439](https://github.com/thanos-io/thanos/pull/5439) Mixin: Add Alert ThanosQueryOverload to Mixin.
+- [#5561](https://github.com/thanos-io/thanos/pull/5561) Query Frontend: Support instant query sharding.
```
Comment (yeya24) [inline CHANGELOG.md:37]: This is added for previous pr.

File: internal/cortex/querier/queryrange/queryrange.proto [hunk 2]
```diff
@@ -66,13 +66,19 @@ message PrometheusInstantQueryData {
 
 message PrometheusInstantQueryResult {
   oneof result {
-    cortexpb.Sample sample = 1;
-    Vector samples = 2;
+    cortexpb.Sample scalar = 1;
+    StringSample stringSample = 2;
+    Vector vector = 3;
+    Matrix matrix = 4;
```
Comment (yeya24) [inline internal/cortex/querier/queryrange/queryrange.proto:72]: Handled all 4 cases mentioned in https://prometheus.io/docs/prometheus/latest/querying/api/#instant-queries.

**Review by fpetkovski — COMMENTED** [review]
I remember running an instant query with a range selector and getting an error, so something like `http_requests_total[2m]` or `http_requests_total[5m:1m]`. Can we cover this case with a unit test?

**Comment by yeya24** [comment]
Sure, I will add that.

**Comment by yeya24** [comment]
cc @fpetkovski @GiedriusS for another look 😄

**Review by fpetkovski — APPROVED** [review]
Thanks!

**Comment by yeya24** [comment]
Need another approval from maintainers and then we can start cutting the rc.

**Review by GiedriusS — APPROVED** [review]

**Comment by GiedriusS** [comment]
:muscle:

**Pull request merged**